File: src/main/java/dev/psygamer/wireframe/item/ItemAttributes.java
package dev.psygamer.wireframe.item;

import dev.psygamer.wireframe.internal.item.InternalItemAttributes;
import dev.psygamer.wireframe.item.util.Rarity;

import net.minecraft.item.Item;
import net.minecraft.item.ItemGroup;

public class ItemAttributes {
	
	protected InternalItemAttributes internal;
	
	protected int maxStackSize = 64;
	
	protected boolean fireResistant = false;
	protected boolean repairable = false;
	
	protected ItemGroup itemGroup;
	protected Item craftingRemainder;
	
	protected Rarity rarity = Rarity.COMMON;
	
	/**
	 * @author psyGamer
	 * @version 1.0 | Minecraft 1.16 +
	 * @since 1.0 | Minecraft 1.16
	 */
	public ItemAttributes() {
		this(null);
	}
	
	/**
	 * @param itemGroup Creative Tab for the Block
	 *
	 * @author psyGamer
	 * @version 1.0 | Minecraft 1.16 +
	 * @since 1.0 | Minecraft 1.16
	 */
	public ItemAttributes(final ItemGroup itemGroup) {
		this.itemGroup = itemGroup;
		
		this.internal = new InternalItemAttributes(this);
	}
	
	public ItemAttributes maxStackSize(final int maxStackSize) {
		this.maxStackSize = maxStackSize;
		
		return this;
	}
	
	public ItemAttributes fireResistant(final boolean fireResistant) {
		this.fireResistant = fireResistant;
		
		return this;
	}
	
	public ItemAttributes repairable(final boolean repairable) {
		this.repairable = repairable;
		
		return this;
	}
	
	public ItemAttributes craftingRemainder(final Item craftingRemainder) {
		this.craftingRemainder = craftingRemainder;
		
		return this;
	}
	
	public ItemAttributes rarity(final Rarity rarity) {
		this.rarity = rarity;
		
		return this;
	}
	
	public int getMaxStackSize() {
		return this.maxStackSize;
	}
	
	public boolean isFireResistant() {
		return this.fireResistant;
	}
	
	public boolean isRepairable() {
		return this.repairable;
	}
	
	public ItemGroup getItemGroup() {
		return this.itemGroup;
	}
	
	public Item getCraftingRemainder() {
		return this.craftingRemainder;
	}
	
	public Rarity getRarity() {
		return this.rarity;
	}
	
	public InternalItemAttributes getInternal() {
		return this.internal;
	}
}

File: src/main/kotlin/dev/psygamer/wireframe/item/ItemAttributes.kt
package dev.psygamer.wireframe.item

import dev.psygamer.wireframe.internal.item.InternalItemAttributes
import dev.psygamer.wireframe.item.util.Rarity
import net.minecraft.item.Item
import net.minecraft.item.ItemGroup

/**
 * A factory class for easily creating attributes related to items.
 *
 * @param itemGroup Creative Tab for the Block
 * @see Item
 */
class ItemAttributes @JvmOverloads constructor(var itemGroup: ItemGroup? = null) {
	
	var maxStackSize = 64
		private set
	var isFireResistant = false
		private set
	var isRepairable = false
		private set
	var craftingRemainder: Item? = null
		private set
	var rarity = Rarity.COMMON
		private set
	
	internal val internal: InternalItemAttributes = InternalItemAttributes(this)
	
	fun maxStackSize(maxStackSize: Int): ItemAttributes {
		this.maxStackSize = maxStackSize
		return this
	}
	
	fun fireResistant(fireResistant: Boolean): ItemAttributes {
		isFireResistant = fireResistant
		return this
	}
	
	fun repairable(repairable: Boolean): ItemAttributes {
		isRepairable = repairable
		return this
	}
	
	fun craftingRemainder(craftingRemainder: Item?): ItemAttributes {
		this.craftingRemainder = craftingRemainder
		return this
	}
	
	fun rarity(rarity: Rarity): ItemAttributes {
		this.rarity = rarity
		return this
	}
}